### PR TITLE
DOC: Document fly, monitor.

### PR DESCRIPTION
**Pull request by danielballan**

### CODE REVIEW

File: doc/source/async.rst [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+Asynchronous Acquistion
+=======================
+
+This section encompasses "fly scans," "monitoring," and in general handling
+data acqusition that is occurring at different rates.
+
+.. note::
+
+    If you are here because you just want to "move two motors at once" or
+    something in that category, you're in luck: your problem is simple! Read
```
Comment (tacaswell) [inline doc/source/async.rst]: 50/50 on quibling over the word 'simple'

File: doc/source/async.rst [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    about multidimensional plans in the section on :doc:`plans`. You don't need
+    anything in this section.
+
+Jargon
+------
+
+**Flying** means: "Let the hardware take control, cache data externally, and
+then transfer all the data to the RunEngine at the end." This is essential when
+the data acquisition rates are faster than the RunEngine or Python can go. As a
+point of reference, the RunEngine processes a message in one second plus any
```
Comment (tacaswell) [inline doc/source/async.rst]: This timing is wrong, It is more like 35k/s

```
In [47]: %timeit RE(Msg('null') for j in range(1000))
10 loops, best of 3: 26.8 ms per loop
```

File: doc/source/async.rst [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+   the RunEngine and processed like any other data.
+
+To "fly" one or more "flyable" devices during a plan, bluesky provides a
+`preprocessor <preprocessors>`. It is available as a wrapper,
+:func:`fly_during_wrapper`
+
+.. code-block:: python
+
+    from bluesky.examples import det
+    # TODO import flyers
```
Comment (klauer) [inline doc/source/async.rst]: `from bluesky.examples import flyer1, flyer2` ?

File: doc/source/async.rst [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    gs.FLYERS = [flyer1, flyer2]
+
+They will be included with all plans until removed.
+
+Monitoring
+----------
+
+To monitor some device during a plan, bluesky provides a
+`preprocessor <preprcoessors>`. It
```
Comment (klauer) [inline doc/source/async.rst]: Link typo

File: bluesky/examples.py [hunk 5]
```diff
@@ -342,32 +410,30 @@ def describe_collect(self):
...
-        self._future.add_done_callback(lambda x: self._finish())
+        st = SimpleStatus()
+        self._completion_status = st
+        self._future.add_done_callback(lambda x: st._finished())
 
         return NullStatus()
 
     def collect(self):
-        if not self.ready:
+        if self._completion_status is not None:
```
Comment (tacaswell) [inline bluesky/examples.py]: This is just checking that the completion object exits, not that it is done.

File: bluesky/examples.py [hunk 6]
```diff
@@ -342,32 +410,30 @@ def describe_collect(self):
...
 
-        # setup the status object (self) that will be returned by
+        # Setup a status object that will be returned by
         # self.complete(). Separately, make dummy status object
         # that is immediately done, and return that, indicated that
         # the 'kickoff' step is done.
         self._future = self.loop.run_in_executor(None, self._scan)
-        self._future.add_done_callback(lambda x: self._finish())
+        st = SimpleStatus()
+        self._completion_status = st
```
Comment (tacaswell) [inline bluesky/examples.py]: This should have some checking that it is not already running.  And probably locking.

File: bluesky/tests/test_examples.py [hunk 7]
```diff
@@ -248,8 +248,8 @@ def callback(name, doc):
 
 def test_simple_fly(fresh_RE):
     RE = fresh_RE
-    mm = MockFlyer(det, motor, RE.loop)
-    RE(fly_gen(mm, -1, 1, 15))
+    mm = MockFlyer('wheeeeee', det, motor, -1, 1, 15, RE.loop)
```
Comment (tacaswell) [inline bluesky/tests/test_examples.py]: :+1: for name

File: bluesky/examples.py (hunk 5 above)
Comment (danielballan) [inline bluesky/examples.py]: When the status is done, the _completion_status attribute is set back to None.

Comment (tacaswell) [inline bluesky/examples.py]: ah, fair enough.

:sheep:

**Comment by danielballan** [comment]
New commits since review doc changes + adding instances of MockFlyer to use in examples + a __repr__. Travis is happy. Self-merging.

**Pull request merged**